File: bridge/RemoteBrowserLauncher.ts
import {
  BrowserLauncher,
  resolveCdpWebSocketUrl,
  type BrowserLaunchOptions,
  type LaunchedBrowser,
} from "./BrowserLauncher.js";

export class RemoteBrowserLauncher extends BrowserLauncher {
  constructor(options: BrowserLaunchOptions = {}, cdp_url: string | null = null) {
    super({ ...options, cdp_url });
  }

  async launch(options: BrowserLaunchOptions = {}): Promise<LaunchedBrowser> {
    const endpoint = options.ws_url ?? options.cdp_url ?? this.options.ws_url ?? this.options.cdp_url;
    if (!endpoint) throw new Error("launch.mode=remote requires upstream.ws_url or cdp_url.");
    const ws_url = await resolveCdpWebSocketUrl(endpoint, "remote cdp_url");
    this.launched = { cdp_url: endpoint, ws_url, close: async () => {} };
    return this.launched;
  }
}

File: client/python/modcdp/RemoteBrowserLauncher.py
from __future__ import annotations

from .BrowserLauncher import BrowserLaunchOptions, BrowserLauncher, LaunchedBrowser


class RemoteBrowserLauncher(BrowserLauncher):
    def __init__(self, options: BrowserLaunchOptions | None = None, cdp_url: str | None = None) -> None:
        super().__init__(options)
        self.cdp_url = cdp_url

    def launch(self, options: BrowserLaunchOptions | None = None) -> LaunchedBrowser:
        if not self.cdp_url:
            raise RuntimeError("launch.mode='remote' requires upstream.ws_url.")
        return {"cdp_url": self.cdp_url, "ws_url": self.cdp_url, "close": lambda: None}
